unit_test/environment_test.py

```diff
@@ -41,9 +41,15 @@ def test_shell_eval():
 def test_shell_eval_and_env():
     environment_recipe = parse_environment('VAR="$(echo "$PREV_VAR" string)"')
 
-    environment_dict = environment_recipe.as_dictionary(prev_environment={"PREV_VAR": "1 2 3"})
+    env = {"PREV_VAR": "1 2 3"}
+    # The above uses `echo` binary which, on Windows, is only shipped by msys2
+    # that ships with Git. This binary just hangs when USERPROFILE is missing.
+    if "USERPROFILE" in os.environ:
+        env["USERPROFILE"] = os.environ["USERPROFILE"]
 
-    assert environment_dict == {"PREV_VAR": "1 2 3", "VAR": "1 2 3 string"}
+    environment_dict = environment_recipe.as_dictionary(prev_environment=env)
+
+    assert environment_dict == {**env, "VAR": "1 2 3 string"}
```

Comment:
Ugh. So yeah, to me it feels like we should just fix the dependency on msys2. Since `echo` is built-in to `cmd`, can we use that instead? I tried this on my Windows PC:

```
joeri@JOERICK-PC C:\Users\joeri>python -c "import subprocess; subprocess.run(['echo', 'something'])"              
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.10_3.10.1520.0_x64__qbz5n2kfra8p0\lib\subprocess.py", line 501, in run
    with Popen(*popenargs, **kwargs) as process:
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.10_3.10.1520.0_x64__qbz5n2kfra8p0\lib\subprocess.py", line 969, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.10_3.10.1520.0_x64__qbz5n2kfra8p0\lib\subprocess.py", line 1438, in _execute_child
    hp, ht, pid, tid = _winapi.CreateProcess(executable, args,
FileNotFoundError: [WinError 2] The system cannot find the file specified

joeri@JOERICK-PC C:\Users\joeri>python -c "import subprocess; subprocess.run(['cmd', '/c', 'echo', 'something'])"
something
```

So `echo` isn't installed, but `cmd /c echo` does work.

I think I'd slightly prefer that to a `python -c "import os;..."` because invoking Python here makes things more complicated again and might not properly test the shell-like behaviours.

Reply:
> Ugh. So yeah, to me it feels like we should just fix the dependency on msys2. Since `echo` is built-in to `cmd`, can we use that instead?

Probably but then we would need to run different command depending on platform which seems to somewhat defeat the point of the cross-platform evaluator of these expressions? `python -c` with os.environ and print would at least work the same on all platforms. There's also some case to be made here that maybe `echo` should be special-cased by cibuildwheel's evaluator and treat it more similarly to how shells treat it - as a builtin command. But that would add to the evaluator's complexity and require a separate PR so I don't know...

Reply:
So to summarise, the options are:

1. patch the test so USERPROFILE is included in the env
1. patch the test so that it uses `cmd /c echo` on windows
1. modify the bash evaluator to add 'echo' as a built-in command

My slight preference is still (2) I think. While we've found on CIrrusCI that the `echo` used is from msys2, so the USERPROFILE fix works there, we haven't needed that fix elsewhere. So I'm guessing that the other platforms have different `echo`s available on path. However, all platforms should have the same `cmd`, since that's part of windows, so that would bring uniformity to this test.

Yes, (3) is also an option, but there are lots of shell builtins, but generally they're backed by a system utility -  I'm not aware of this being a problem in practise. So let's keep that in the back pocket - if there's a bug report about missing `echo`, we can implement a builtin.

Reply:
> While we've found on CIrrusCI that the `echo` used is from msys2, so the USERPROFILE fix works there, we haven't needed that fix elsewhere.

I'm reasonably sure that all Windows runners use `echo` from Git/msys2 but for some reason the bug is only reproducible on Cirrus CI (maybe it's specific to the version of Git/msys2?). I was not able to reproduce the hang locally in a Windows VM (without Git it does fail due to `FileNotFoundError` as `echo` is indeed missing) which would explain why it wouldn't necessarily be reproducible in all other CIs, even if they use `echo` from msys2 (which I would imagine they have to unless there's some other external program that provides `echo` on Windows).

Reply:
Either way, patching the test to use `cmd /c echo` on windows is fine I guess, I can make that change.

Reply:
Resolved by #1196.